### ISSUE TITLE
Refactor `user-input-settings` REST route and preload it

### ISSUE THREAD

**Issue by jimmymadon** [google/site-kit-wp#6233]
## Feature Description

Site Kit’s current internal REST endpoint to fetch and save responses for the User Input questionnaire is `core/user/data/user-input-settings` within the `Core\REST_API\REST_Routes` class. This route should be refactored into the `User_Input_Settings` class. This will keep the `REST_Routes` class generic and will register routes defined in individual classes via the `googlesitekit_rest_routes` filter.

Refer to the [relevant section](https://docs.google.com/document/d/1k8NOo2xEHUQAcgYG0TB5sUjGcPiC8SWIalEdd9G8-9g/edit?pli=1#heading=h.3c9jhdvp9ikx) in the design doc for further context.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* ~~The `core/user/data/user-input-settings` REST route should be moved to the `User_Input_Settings` class.~~ **This was already moved to an appropriate controller class in #5898**
* The above route should also be pre-fetched, i.e. added to the `googlesitekit_apifetch_preload_paths` filter.

## Implementation Brief

- The `core/user/data/user-input-settings` REST endpoint is already being moved to `User_Input` as a part of #5898. Please wait until it is merged.
- In `includes/Core/User_Input/REST_User_Input_Controller.php` (being implemented in #5898):
    - In the `register()` method, include the `/core/user/data/user-input-settings` endpoint in the pre-fetched API responses with the `googlesitekit_apifetch_preload_paths` filter. An example of this can be found in `includes/Core/Dismissals/REST_Dismissals_Controller.php`.

### Test Coverage

- In `tests/phpunit/integration/Core/User_Input/REST_User_Input_ControllerTest.php` (being implemented in #5898):
    - Update the `test_register()` test case to include assertions for the `googlesitekit_apifetch_preload_paths` filter. Example can be found in `tests/phpunit/integration/Core/Dismissals/REST_Dismissals_ControllerTest.php`.

## QA Brief

* Should not introduce any visual changes to end user.
* Verify the User Input Questions still work as expected.
  * Verify answering the User Input Questions works as expected.
  * Verify editing the User Input Questions at `Settings > Admin Settings` works as expected.
* This issue pre-loads answers to the User Input questions when Site Kit loads, i.e. not requiring a network request on-demand when loading the user input screen. Go to `/wp-admin/admin.php?page=googlesitekit-user-input` and verify that an initial `GET` request to `/wp-json/google-site-kit/v1/core/user/data/user-input-settings` is NOT made, which is currently present in the latest release.

## Changelog entry

* Preload `user-input-settings` REST route.

**Comment by tofumatt** [comment]
ACs ✅ 

Moving to IB 👍🏻

**Comment by nfmohit** [comment]
**Note to IB Reviewer:** The `core/user/data/user-input-settings` REST endpoint is already being moved to `User_Input` as a part of #5898 (see [this file](https://github.com/google/site-kit-wp/blob/2001755b417ba12a1905a3c41cb9e4d098924dc0/includes/Core/User_Input/REST_User_Input_Controller.php)). As a result, I have set #5898 as a blocker for this issue.

**Comment by techanvil** [comment]
IB :white_check_mark:

**Comment by aaemnnosttv** [comment]
Noting here that I've amended the AC to be consistent with _not moving_ the route definition from where it is now (in its controller) where it should be (but wasn't when this issue was defined).

This issue is essentially just about the remaining part about preloading the request which is still valid.

**Comment by mohitwp** [comment]
### QA Update ⚠️ 

Verified on dev.

@derweili 
I've noticed that under admin>settings, changes to answers is not getting save on develop. Edit flow is also changed on develop, so I'm not sure if it is due to connected PR or any other PR. Can you please confirm if this issue is related to this PR or not ?


https://user-images.githubusercontent.com/94359491/211258442-7fd2841a-5e03-4258-a4c9-c1ce9aae666f.mp4

**Comment by derweili** [comment]
@mohitwp can confirm this behaviour.

Actually I do not know how this is supposed to work. Should it save automatically in the background? Is there a "Save changes" button missing?

However, I verified the with the commit b998f2237c1be7b63ec0e654880c30ebe663f8c7 which is the commit on develop branch before the PR of this issue was merged. The error/behaviour appears there, too.
So I think, the error is not related to this issue.

**Comment by mohitwp** [comment]
Thank you, @derweili. I will create a separate issue for this.

**Comment by nfmohit** [comment]
> I've noticed that under admin>settings, changes to answers is not getting save on develop. Edit flow is also changed on develop, so I'm not sure if it is due to connected PR or any other PR.

@mohitwp I believe this came up after #5891 was merged, where we introduced inline editing to the User Input Preview screen. This may sound confusing because Admin Settings and User Input Preview are different screens, but we actually re-use the User Input Preview component in the Admin Settings.

The ability to save the answers in Admin Settings will be restored as a part of #5897.

**Comment by mohitwp** [comment]
Thank you @nfmohit 

### QA Update ✅ 

- Verified on dev.
- Network request on-demand when loading the user input screen is not getting execute on dev branch.
- Verified the user input flow.
- Issue> Changes in answers not getting save under Admin settings may be due to changes introduced in https://github.com/google/site-kit-wp/issues/5891 and will get resolve through https://github.com/google/site-kit-wp/issues/5891.
- Tested user input flow, no visual changes found.

<details>

Develop -

![image](https://user-images.githubusercontent.com/94359491/211296084-f62dd042-21a3-4a1b-aa2b-f74ae992c8c4.png)

Latest -

![image](https://user-images.githubusercontent.com/94359491/211296126-ed0db142-104a-4666-9094-ac054047b748.png)

</details>

**Comment by felixarntz** [comment]
## Approval ❌ 

@derweili @aaemnnosttv @techanvil The endpoint is now preloaded regardless of whether the `userInput` feature flag is enabled or not. This consideration wasn't called out in the ACs, but it's a problem that has to be fixed.

I'm wondering why we even load all that user input-specific code (e.g. the entire `User_Input` class and everything else from there) if the flag is not enabled. But at a minimum we will have to wrap this preload logic to only preload if the flag is enabled.

**Comment by aaemnnosttv** [comment]
> I'm wondering why we even load all that user input-specific code (e.g. the entire `User_Input` class and everything else from there) if the flag is not enabled.

There was an issue for this https://github.com/google/site-kit-wp/issues/6347 but this is only in develop. Does that address your concern or did you mean something else?

**Comment by felixarntz** [comment]
@aaemnnosttv The preloading of the endpoint when the feature flag isn't active is what concretely needs to be fixed as otherwise it's a regression introduced here. Squeezing in #6347 is probably too much, so then we should go with the approach of checking for the feature flag in the preload filter.

**Comment by mohitwp** [comment]
QA Update ✅ 

- Verified on main.
- Now the endpoint is not preloading when the feature flag isn't active.
- Tested on main, entity, WP Dashboard and SK settings page.
- Done quick overview of User Input questions flow.
- Verified main and entity dashboard when 'User Input' feature flag is not active.
- I'm getting Site Kit API error in console when 'User Input' feature flag is active. I will create a separate issue for this also.

<details>

![image](https://user-images.githubusercontent.com/94359491/212330389-d11be9b0-c4c2-46d0-b105-70e9b7674979.png)

</details>

When Feature flag is enabled

![image](https://user-images.githubusercontent.com/94359491/212362282-f917b07f-0e89-4412-af87-c20dd730bbfb.png)

When Feature flag is not enabled.

![image](https://user-images.githubusercontent.com/94359491/212362420-5963bae7-52c1-42db-8ac5-d56561008ac9.png)

**Comment by felixarntz** [comment]
Thanks all for the quick fix!
